pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -468,10 +468,13 @@ private void doBatchSendAndAdd(MessageImpl<T> msg, SendCallback callback, ByteBu
             log.debug("[{}] [{}] Closing out batch to accommodate large message with size {}", topic, producerName,
                     msg.getDataBuffer().readableBytes());
         }
-        batchMessageAndSend();
-        batchMessageContainer.add(msg, callback);
-        lastSendFuture = callback.getFuture();
-        payload.release();
+        try {
+            batchMessageAndSend();
+            batchMessageContainer.add(msg, callback);
+            lastSendFuture = callback.getFuture();
+        } finally {
```

Comment:
If we do get exception here, shouldn't we also trigger the send callback?

Reply:
yes, we trigger the callback but we missed to reduce to refCnt of `payload` in producer so, when [callback receives callback](https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L257), it tries to release payload with refCnt=2 so, producer doesn't release the actual payload.